crates/matrix-sdk/src/event_cache/deduplicator.rs

```diff
@@ -0,0 +1,266 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Simple but efficient types to find duplicated events. See [`Deduplicator`]
+//! to learn more.
+
+use std::{collections::BTreeSet, sync::Mutex};
+
+use growable_bloom_filter::{GrowableBloom, GrowableBloomBuilder};
+
+use super::store::{Event, RoomEvents};
+
+/// `Deduplicator` is an efficient type to find duplicated events.
+///
+/// It uses a [bloom filter] to provide a fast probabilistic answer to: “has
+/// event E been seen already?”. False positives are possible, while false
+/// negatives are impossible. In the case of a positive reply, we fallback to a
+/// linear (backward) search on all events to check whether it's a false
+/// positive or not
+///
+/// [bloom filter]: https://en.wikipedia.org/wiki/Bloom_filter
+pub struct Deduplicator {
+    bloom_filter: Mutex<GrowableBloom>,
+}
+
+impl Deduplicator {
+    const APPROXIMATED_MAXIMUM_NUMBER_OF_EVENTS: usize = 800_000;
+    const DESIRED_FALSE_POSITIVE_RATE: f64 = 0.001;
+
+    /// Create a new `Deduplicator`.
+    pub fn new() -> Self {
+        Self {
+            bloom_filter: Mutex::new(
+                GrowableBloomBuilder::new()
+                    .estimated_insertions(Self::APPROXIMATED_MAXIMUM_NUMBER_OF_EVENTS)
+                    .desired_error_ratio(Self::DESIRED_FALSE_POSITIVE_RATE)
+                    .build(),
+            ),
+        }
+    }
+
+    /// Scan a collection of events and detect duplications.
+    ///
+    /// This method takes a collection of events `events_to_scan` and returns a
+    /// new collection of events, where each event is decorated by a
+    /// [`Decoration`], so that the caller can decide what to do with these
+    /// events.
+    ///
+    /// Each scanned event will update `Self`'s internal state.
+    ///
+    /// `existing_events` represents all events of a room that already exist.
+    pub fn scan_and_learn<'a, I>(
+        &'a self,
+        events_to_scan: I,
+        existing_events: &'a RoomEvents,
+    ) -> impl Iterator<Item = Decoration<I::Item>> + 'a
+    where
+        I: Iterator<Item = Event> + 'a,
```

Comment:
I'm not sure about this API's signature: instead of taking an iterator and returning one, could we have a function that operates on a single item at a time, returning a decoration for it? `already_seen` could be a member of this struct (which would be more optimized too), and with respect to the YAGNI about the bloom filter, we could likely only use `already_seen`…